[DependencyInjection] Define default priority inside service class

pcabreus opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 4.4 for features
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no     
| Deprecations? | no
| Tests pass?   | yes   
| Fixed tickets | #...   
| License       | MIT

With autoconfigure functions we don't have to do much to inject tagged services into a collector using interfaces, using `!tagged` is the easy way. But if we need some order we have to use priority, the problem is that priority only can be set by configuration file when it is a simple value.
This PR take a look if the static method `getDefaultPriority` exists in the service class and return its default value for the service instead of current 0 by defualt.

Example of a service with default priority:

    <?php
    namespace App;
    use App\ServiceTagged;
    class Service implement ServiceTagged
    {
        public static function getDefaultPriority(): int
        {
            return 50; //This will be the default priority if it's not defined in a configuration file
        }
    }
    ?>

PD: First PR please by nice!

pcabreus committed:
Define default priority inside service class

With autoconfigure functions we don't have to do anything to inject tagged services into a collector using interfaces to define tags. But if we need some order we have to use priority, the problem is that priority only can be set by configuration file when it is a simple value.
With this change you could instead of put 0 by defualt take a look if the static method `getDefaultPriority` exists in the service class and return its default value for the service.

Example:

    <?php
    namespace App;
    use App\ServiceTagged;
    class Service implement ServiceTagged
    {
        public static function getDefaultPriority(): int
        {
            return 50; //This will be the default priority if it's not defined in a configuration file
        }
    }
    ?>

yceruto commented on src/Symfony/Component/DependencyInjection/Compiler/PriorityTaggedServiceTrait.php:
```diff
@@ -53,7 +53,7 @@ private function findAndSortTaggedServices($tagName, ContainerBuilder $container
 
         foreach ($container->findTaggedServiceIds($tagName, true) as $serviceId => $attributes) {
-            $priority = isset($attributes[0]['priority']) ? $attributes[0]['priority'] : 0;
+            $priority = isset($attributes[0]['priority']) ? $attributes[0]['priority'] : (method_exists($serviceId, 'getDefaultPriority') ? $serviceId::getDefaultPriority() : 0);
```
* The service id is not necessarily the FQCN; you should use the class of its definition instead (e.g. `$container->getDefinition($serviceId)->getClass()`).
* Looking for an ambitrary method can cause unexpected behaviors (i.e., what if my service uses this method for another purpose?) An alternative may be to add a new interface that forces this contract and look for this interface instead.

nicolas-grekas replied:
The class is already extracted a few lines below, this should be reused.
Not sure about the interface Yonel, that wouldn't make much sense from the class' design pov I think.
Instead, I'd suggest supporting this only when a TaggedIteratorArgument is used, and add a getDefaultPriorityMethod on it. The default implementation would return `getDefaultMyTagPriority`, exactly as `getDefaultIndexMethod` returns `getDefaultMyTagName`

pcabreus replied:
I used `getClass` on my own implementation without using this trait and yes we could use `$class` extracted.
I wouldn't like the new interface, maybe a good solution is do the same of `getDefailtIndexMethod` like Nicolas says.

vudaltsov replied:
@nicolas-grekas , am I right that what you suggest will be an expansion of https://github.com/symfony/symfony/pull/30257?

nicolas-grekas replied:
Absolutely

yceruto reviewed (commented):
Hola! Congratulations for your first attempt :) :tada:

pcabreus committed:
Update YamlFileLoader.php

It looks like `YamlFileLoader` makes `index_by` mandatory when you define the tagged service using `arguments: [!tagged { tag: 'app.handler'}]` syntax. It's needed to accept others parameters like the new `default_priority_method` without specify `index_by` and the `XmlFielLoader` seems to accept it.

pcabreus committed:
added ability to define an priority method for a tagged collection

pcabreus committed:
Merge branch '4.4' into patch-1

pcabreus committed:
fixed fabbot.io code standard

yceruto commented on src/Symfony/Component/DependencyInjection/Compiler/PriorityTaggedServiceTrait.php:
```diff
@@ -41,39 +41,82 @@ trait PriorityTaggedServiceTrait
+            $priority = 0;
+            if ($defaultPriorityMethod && $r->hasMethod($defaultPriorityMethod)) {
+                if (!($rpm = $r->getMethod($defaultPriorityMethod))->isStatic()) {
+                    throw new InvalidArgumentException(
```
should be in one line (same bellow)

pcabreus replied:
That was an IDE decision, I'll fix it

yceruto commented on src/Symfony/Component/DependencyInjection/Loader/Configurator/ContainerConfigurator.php:
```diff
@@ -118,31 +118,48 @@ function iterator(array $values): IteratorArgument
+ * @param string|null $defaultIndexMethod
+ * @param bool $needsIndexes
+ * @param string|null $defaultPriorityMethod
+ * @return TaggedIteratorArgument
```
should be reverted unless you need to add some useful description (same bellow)

pcabreus replied:
Same I'll fix it

yceruto reviewed (commented):
Others code standard that aren't cover by fabbot.io:

vudaltsov commented on src/Symfony/Component/DependencyInjection/CHANGELOG.md:
```diff
@@ -6,6 +6,7 @@ CHANGELOG
 
  * deprecated support for short factories and short configurators in Yaml
  * deprecated `tagged` in favor of `tagged_iterator`
+ * added ability to define an priority method for a tagged collection
```
```suggestion
 * added ability to define a priority method for a tagged collection
```

vudaltsov commented on src/Symfony/Component/DependencyInjection/Compiler/PriorityTaggedServiceTrait.php:
```diff
@@ -41,39 +41,82 @@ trait PriorityTaggedServiceTrait
+            $class = $container->getDefinition($serviceId)->getClass();
+            $class = $container->getParameterBag()->resolveValue($class) ?: null;
+
+            if (!$r = $container->getReflectionClass($class)) {
```
```suggestion
            if (null !== $r = $container->getReflectionClass($class)) {
```

vudaltsov commented on src/Symfony/Component/DependencyInjection/Compiler/PriorityTaggedServiceTrait.php:
```diff
@@ -41,39 +41,82 @@ trait PriorityTaggedServiceTrait
+
+            $priority = 0;
+            if ($defaultPriorityMethod && $r->hasMethod($defaultPriorityMethod)) {
+                if (!($rpm = $r->getMethod($defaultPriorityMethod))->isStatic()) {
```
Let's use a more comprehensible variable name here: `$priorityReflMethod`

pcabreus replied:
Agree

vudaltsov commented on src/Symfony/Component/DependencyInjection/Compiler/PriorityTaggedServiceTrait.php:
```diff
@@ -41,39 +41,82 @@ trait PriorityTaggedServiceTrait
+                if (!($rpm = $r->getMethod($defaultPriorityMethod))->isStatic()) {
+                    throw new InvalidArgumentException(
+                        sprintf(
+                            'Method "%s::%s()" should be static: tag "%s" on service "%s" is missing default priority method.',
```
I think that `tag ... is missing` is not actually correct. It's not **missing** it, since the method does exist, it's just not accessible. What if we try smth like this: 
```suggestion
                            'Default priority method "%s::%s()" of the "%s"-tagged collection on service "%s" must be static.',
```

pcabreus committed:
fixed spell

Co-Authored-By: Valentin Udaltsov <udaltsov.valentin@gmail.com>

pcabreus committed:
compare using null

Co-Authored-By: Valentin Udaltsov <udaltsov.valentin@gmail.com>

pcabreus committed:
better exception decription

Co-Authored-By: Valentin Udaltsov <udaltsov.valentin@gmail.com>

pcabreus committed:
fixed some code standard

pcabreus committed:
fixed some code standard

kevin-verschaeve commented on src/Symfony/Component/DependencyInjection/Compiler/PriorityTaggedServiceTrait.php:
```diff
@@ -41,39 +41,57 @@ trait PriorityTaggedServiceTrait
+                $priority = $priorityReflMethod->invoke(null);
+
+                if (!\is_int($priority)) {
+                    throw new InvalidArgumentException(sprintf('Default priority method "%s::%s()" of the "%s"-tagged collection on service "%s" should return an integer, got %s.', $class, $defaultPriorityMethod, \gettype($priority), $tagName, $serviceId));
```
are arguments in the correct order ? I would write :

```suggestion
                    throw new InvalidArgumentException(sprintf('Default priority method "%s::%s()" of the "%s"-tagged collection on service "%s" should return an integer, got %s.', $class, $defaultPriorityMethod, $tagName, $serviceId, \gettype($priority)));
```

pcabreus committed:
fixed arguments order

Co-Authored-By: Kevin Verschaeve <kevin.verschaeve@live.fr>

pcabreus committed:
using the `tagged_iterator` instead deprecated `tagged`

pcabreus committed:
optional priority method unless you define it

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Compiler/PriorityTaggedServiceTrait.php:
```diff
@@ -41,42 +41,61 @@ trait PriorityTaggedServiceTrait
             }
 
-            $class = $r->name;
+            $class = $reflectionClass->name;
```
let's keep `$r` as the name here, that will reduce the diff, thus reduce future merge conflicts

nicolas-grekas reviewed (commented):
Thanks and sorry for the late review.
Here are some comments. Please rebase also.
Looks good to me otherwise.

nicolas-grekas commented:
Wait, this has already been implemented in #33628!
Thanks for pushing this forward and sorry for the confusion.
